File: matcher.py
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass, field
from pathlib import Path
from typing import Iterable

from difflib import SequenceMatcher
from typing import Any

from members import Member
from storage import AliasStore


DAY_ORDER = ["monday", "tuesday", "wednesday", "thursday", "friday", "saturday"]


def normalize_name(value: str) -> str:
    text = unicodedata.normalize("NFKC", value).casefold()
    text = text.replace("ı", "i")
    # Keep letters/numbers across scripts; remove decorative punctuation/spacing.
    text = "".join(ch for ch in text if ch.isalnum())
    return text


@dataclass(slots=True)
class Observation:
    source_file: str
    day: str
    rank: int
    raw_player_id: int | None
    raw_name: str
    points: int
    extraction_confidence: float
    is_pinned_row: bool
    matched_member_id: int | None = None
    matched_member_name: str | None = None
    match_method: str = "unmatched"
    match_confidence: float = 0.0
    issue: str | None = None
    alternatives: list[tuple[int, str, float]] = field(default_factory=list)


@dataclass(slots=True)
class WeeklyData:
    observations: list[Observation]
    scores: dict[tuple[int, str], int]
    issues: list[str]
    missing_by_day: dict[str, list[Member]]


class MemberMatcher:
    def __init__(self, members: Iterable[Member], alias_store: AliasStore):
        self.members = list(members)
        self.alias_store = alias_store
        self.by_id: dict[int, list[Member]] = {}
        for m in self.members:
            self.by_id.setdefault(m.member_id, []).append(m)
        self.by_norm: dict[str, list[Member]] = {}
        for m in self.members:
            self.by_norm.setdefault(normalize_name(m.name), []).append(m)
        self.name_choices = {m.member_id: normalize_name(m.name) for m in self.members}

    def _set_match(self, obs: Observation, member: Member, method: str, confidence: float) -> Observation:
        obs.matched_member_id = member.member_id
        obs.matched_member_name = member.name
        obs.match_method = method
        obs.match_confidence = confidence
        return obs

    def match(self, obs: Observation) -> Observation:
        if obs.raw_player_id is not None:
            candidates = self.by_id.get(obs.raw_player_id, [])
            if len(candidates) == 1:
                return self._set_match(obs, candidates[0], "id", 1.0)
            if len(candidates) > 1:
                obs.issue = f"Duplicate active member ID {obs.raw_player_id}; manual review required."
                return obs

        norm = normalize_name(obs.raw_name)
        exact = self.by_norm.get(norm, [])
        if len(exact) == 1:
            return self._set_match(obs, exact[0], "exact_name", 1.0)
        if len(exact) > 1:
            obs.issue = f"Normalized name {obs.raw_name!r} maps to multiple active members."
            return obs

        alias_member_id = self.alias_store.get_member_id(norm)
        if alias_member_id is not None:
            candidates = self.by_id.get(alias_member_id, [])
            if len(candidates) == 1:
                return self._set_match(obs, candidates[0], "saved_alias", 1.0)

        # Fuzzy matching is suggestion-only. It never auto-assigns an identity.
        if norm:
            scored: list[tuple[float, int]] = []
            for member_id, candidate in self.name_choices.items():
                score = SequenceMatcher(None, norm, candidate).ratio()
                scored.append((score, member_id))
            scored.sort(reverse=True)
            obs.alternatives = []
            for score, member_id in scored[:3]:
                member = self.by_id.get(int(member_id), [None])[0]
                if member is not None:
                    obs.alternatives.append((member.member_id, member.name, score))
        obs.issue = "No deterministic ID/name/alias match."
        return obs

    def manual_assign(self, obs: Observation, member_id: int, remember_alias: bool = True) -> Observation:
        candidates = self.by_id.get(member_id, [])
        if len(candidates) != 1:
            raise ValueError(f"Member ID {member_id} is not uniquely active")
        member = candidates[0]
        self._set_match(obs, member, "manual", 1.0)
        obs.issue = None
        if remember_alias:
            norm = normalize_name(obs.raw_name)
            if norm:
                self.alias_store.save_alias(norm, obs.raw_name, member_id)
        return obs


def observations_from_extractions(results: list[Any]) -> tuple[list[Observation], list[str]]:
    observations: list[Observation] = []
    issues: list[str] = []
    for result in results:
        if result.error:
            issues.append(f"{result.image_path.name}: extraction failed: {result.error}")
            continue
        assert result.extraction is not None
        extraction = result.extraction
        for row in extraction.rows:
            observations.append(_make_observation(result.image_path, extraction.detected_day, row, False))
        if extraction.pinned_row is not None:
            observations.append(_make_observation(result.image_path, extraction.detected_day, extraction.pinned_row, True))
        for warning in extraction.warnings:
            issues.append(f"{result.image_path.name}: model warning: {warning}")
    return observations, issues


def _make_observation(path: Path, day: str, row: Any, pinned: bool) -> Observation:
    return Observation(
        source_file=path.name,
        day=day,
        rank=row.rank,
        raw_player_id=row.player_id,
        raw_name=row.raw_name,
        points=row.points,
        extraction_confidence=row.extraction_confidence,
        is_pinned_row=pinned,
    )


def build_weekly_data(observations: list[Observation], members: list[Member], base_issues: list[str] | None = None) -> WeeklyData:
    issues = list(base_issues or [])
    scores: dict[tuple[int, str], int] = {}
    grouped: dict[tuple[int, str], list[Observation]] = {}

    for obs in observations:
        if obs.matched_member_id is None:
            issues.append(
                f"Unmatched: {obs.day} rank {obs.rank} {obs.raw_name!r} ({obs.points:,}) from {obs.source_file}"
            )
            continue
        grouped.setdefault((obs.matched_member_id, obs.day), []).append(obs)

    for key, group in grouped.items():
        values = sorted({o.points for o in group})
        # Scores appear cumulative. For duplicate observations collected minutes apart,
        # preserve the highest value and surface the disagreement for review/audit.
        scores[key] = max(values)
        if len(values) > 1:
            member_id, day = key
            issues.append(
                f"Conflicting duplicate scores for member {member_id} on {day}: "
                f"{', '.join(f'{v:,}' for v in values)}. Highest value used."
            )

    observed_days = sorted({o.day for o in observations}, key=lambda d: DAY_ORDER.index(d) if d in DAY_ORDER else 99)
    missing_by_day: dict[str, list[Member]] = {}
    for day in observed_days:
        matched_ids = {member_id for member_id, d in scores if d == day}
        missing_by_day[day] = [m for m in members if m.member_id not in matched_ids]

    return WeeklyData(observations, scores, issues, missing_by_day)

File: dotnet/LastWarExtractor/Services/Matcher.cs
using LastWarExtractor.Models;

namespace LastWarExtractor.Services;

/// <summary>
/// Match extracted observations to active members. Two phases: deterministic identity
/// signals (ID, exact name, saved alias), then conservative avatar-assisted matching.
/// Ported from matcher.py.
/// </summary>
public sealed class MemberMatcher
{
    private readonly AliasStore _aliasStore;
    private readonly AvatarStore? _avatarStore;
    private readonly double _avatarAutoThreshold;
    private readonly double _avatarMinMargin;
    private readonly double _avatarSuggestionThreshold;

    private readonly Dictionary<int, List<Member>> _byId = new();
    private readonly Dictionary<string, List<Member>> _byNorm = new();
    private readonly Dictionary<int, string> _nameChoices = new();

    public MemberMatcher(
        IEnumerable<Member> members,
        AliasStore aliasStore,
        AvatarStore? avatarStore = null,
        double avatarAutoThreshold = 0.92,
        double avatarMinMargin = 0.06,
        double avatarSuggestionThreshold = 0.78)
    {
        _aliasStore = aliasStore;
        _avatarStore = avatarStore;
        _avatarAutoThreshold = avatarAutoThreshold;
        _avatarMinMargin = avatarMinMargin;
        _avatarSuggestionThreshold = avatarSuggestionThreshold;

        foreach (var m in members)
        {
            if (!_byId.TryGetValue(m.MemberId, out var idList))
                _byId[m.MemberId] = idList = new List<Member>();
            idList.Add(m);

            var norm = TextUtil.NormalizeName(m.Name);
            if (!_byNorm.TryGetValue(norm, out var normList))
                _byNorm[norm] = normList = new List<Member>();
            normList.Add(m);

            _nameChoices[m.MemberId] = norm;
        }
    }

    public IEnumerable<int> MemberIds => _byId.Keys;

    private static void SetMatch(Observation obs, Member member, string method, double confidence)
    {
        obs.MatchedMemberId = member.MemberId;
        obs.MatchedMemberName = member.Name;
        obs.MatchMethod = method;
        obs.MatchConfidence = confidence;
        obs.Issue = null;
    }

    public void MatchDeterministic(Observation obs)
    {
        if (obs.RawPlayerId is int playerId)
        {
            var candidates = _byId.GetValueOrDefault(playerId) ?? new List<Member>();
            if (candidates.Count == 1)
            {
                SetMatch(obs, candidates[0], "id", 1.0);
                return;
            }
            if (candidates.Count > 1)
            {
                obs.Issue = $"Duplicate active member ID {playerId}; manual review required.";
                return;
            }
        }

        var norm = TextUtil.NormalizeName(obs.RawName);
        var exact = _byNorm.GetValueOrDefault(norm) ?? new List<Member>();
        if (exact.Count == 1)
        {
            SetMatch(obs, exact[0], "exact_name", 1.0);
            return;
        }
        if (exact.Count > 1)
        {
            obs.Issue = $"Normalized name '{obs.RawName}' maps to multiple active members.";
            return;
        }

        var aliasMemberId = _aliasStore.GetMemberId(norm);
        if (aliasMemberId is int aliasId)
        {
            var candidates = _byId.GetValueOrDefault(aliasId) ?? new List<Member>();
            if (candidates.Count == 1)
            {
                SetMatch(obs, candidates[0], "saved_alias", 1.0);
                return;
            }
        }

        SetFuzzySuggestions(obs);
        obs.Issue = "No deterministic ID/name/alias match.";
    }

    private void SetFuzzySuggestions(Observation obs)
    {
        var norm = TextUtil.NormalizeName(obs.RawName);
        var fuzzy = new List<Alternative>();
        if (norm.Length > 0)
        {
            var scored = _nameChoices
                .Select(kv => (Score: TextUtil.SequenceRatio(norm, kv.Value), MemberId: kv.Key))
                .OrderByDescending(x => x.Score)
                .ThenByDescending(x => x.MemberId)
                .Take(3);
            foreach (var (score, memberId) in scored)
            {
                var member = _byId.GetValueOrDefault(memberId)?.FirstOrDefault();
                if (member is not null)
                    fuzzy.Add(new Alternative(member.MemberId, member.Name, score));
            }
        }
        obs.Alternatives = fuzzy;
    }

    public bool LearnAvatar(Observation obs)
    {
        if (_avatarStore is null
            || obs.MatchedMemberId is null
            || string.IsNullOrEmpty(obs.AvatarFingerprint)
            || obs.MatchMethod is not ("id" or "exact_name" or "saved_alias" or "manual" or "avatar_auto"))
            return false;
        return _avatarStore.SaveReference(obs.MatchedMemberId.Value, obs.AvatarFingerprint, obs.SourceFile, obs.MatchMethod);
    }

    public void MatchAvatar(Observation obs)
    {
        if (obs.MatchedMemberId is not null || _avatarStore is null || string.IsNullOrEmpty(obs.AvatarFingerprint))
            return;

        var matches = _avatarStore.BestMatches(obs.AvatarFingerprint!, _byId.Keys, limit: 5);
        if (matches.Count == 0)
            return;

        var best = matches[0];
        double secondScore = matches.Count > 1 ? matches[1].Score : 0.0;
        double margin = best.Score - secondScore;
        var memberCandidates = _byId.GetValueOrDefault(best.MemberId) ?? new List<Member>();

        if (memberCandidates.Count == 1 && best.Score >= _avatarAutoThreshold && margin >= _avatarMinMargin)
        {
            SetMatch(obs, memberCandidates[0], "avatar_auto", best.Score);
            LearnAvatar(obs);
            return;
        }

        var avatarAlternatives = new List<Alternative>();
        foreach (var match in matches)
        {
            if (match.Score < _avatarSuggestionThreshold)
                continue;
            var members = _byId.GetValueOrDefault(match.MemberId) ?? new List<Member>();
            if (members.Count == 1)
                avatarAlternatives.Add(new Alternative(match.MemberId, members[0].Name, match.Score));
        }

        var merged = new Dictionary<int, (string Name, double Score)>();
        foreach (var alt in obs.Alternatives)
            merged[alt.MemberId] = (alt.Name, alt.Score);
        foreach (var alt in avatarAlternatives)
        {
            if (!merged.TryGetValue(alt.MemberId, out var current) || alt.Score > current.Score)
                merged[alt.MemberId] = (alt.Name, alt.Score);
        }
        obs.Alternatives = merged
            .Select(kv => new Alternative(kv.Key, kv.Value.Name, kv.Value.Score))
            .OrderByDescending(a => a.Score)
            .Take(5)
            .ToList();

        if (best.Score >= _avatarSuggestionThreshold)
            obs.Issue = $"Avatar suggests member {best.MemberId} at {best.Score:P0} (margin {margin:P0}); manual review required.";
    }

    public void ManualAssign(Observation obs, int memberId, bool rememberAlias = true)
    {
        var candidates = _byId.GetValueOrDefault(memberId) ?? new List<Member>();
        if (candidates.Count != 1)
            throw new InvalidOperationException($"Member ID {memberId} is not uniquely active");
        SetMatch(obs, candidates[0], "manual", 1.0);
        if (rememberAlias)
        {
            var norm = TextUtil.NormalizeName(obs.RawName);
            if (norm.Length > 0)
                _aliasStore.SaveAlias(norm, obs.RawName, memberId);
        }
        LearnAvatar(obs);
    }
}

public static class ObservationBuilder
{
    public static (List<Observation> Observations, List<string> Issues) FromExtractions(IReadOnlyList<ExtractionResult> results)
    {
        var observations = new List<Observation>();
        var issues = new List<string>();
        foreach (var result in results)
        {
            if (result.Error is not null)
            {
                issues.Add($"{result.FileName}: extraction failed: {result.Error}");
                continue;
            }
            var extraction = result.Extraction!;
            foreach (var row in extraction.Rows)
            {
                var (obs, issue) = MakeObservation(result.ImagePath, extraction.DetectedDay, row, false);
                observations.Add(obs);
                if (issue is not null)
                    issues.Add(issue);
            }
            if (extraction.PinnedRow is not null)
            {
                var (obs, issue) = MakeObservation(result.ImagePath, extraction.DetectedDay, extraction.PinnedRow, true);
                observations.Add(obs);
                if (issue is not null)
                    issues.Add(issue);
            }
            foreach (var warning in extraction.Warnings)
                issues.Add($"{result.FileName}: model warning: {warning}");
        }
        return (observations, issues);
    }

    private static (Observation, string?) MakeObservation(string path, string day, ExtractedRow row, bool pinned)
    {
        (int, int, int, int)? bbox = row.AvatarBBox?.AsTuple();
        var fingerprint = Fingerprinter.FingerprintFromBBox(path, bbox);
        string? issue = null;
        if (bbox is not null && fingerprint is null)
            issue = $"{Path.GetFileName(path)}: could not fingerprint avatar for '{row.RawName}' at rank {row.Rank}.";

        var obs = new Observation
        {
            SourceFile = Path.GetFileName(path),
            Day = day,
            Rank = row.Rank,
            RawPlayerId = row.PlayerId,
            RawName = row.RawName,
            Points = row.Points,
            ExtractionConfidence = row.ExtractionConfidence,
            IsPinnedRow = pinned,
            AvatarBBox = bbox,
            AvatarFingerprint = fingerprint,
        };
        return (obs, issue);
    }
}

public static class WeeklyBuilder
{
    public static WeeklyData Build(List<Observation> observations, List<Member> members, List<string>? baseIssues = null)
    {
        var issues = new List<string>(baseIssues ?? new List<string>());
        var scores = new Dictionary<(int, string), int>();
        var grouped = new Dictionary<(int, string), List<Observation>>();

        foreach (var obs in observations)
        {
            if (obs.MatchedMemberId is null)
            {
                issues.Add($"Unmatched: {obs.Day} rank {obs.Rank} '{obs.RawName}' ({obs.Points:N0}) from {obs.SourceFile}");
                continue;
            }
            var key = (obs.MatchedMemberId.Value, obs.Day);
            if (!grouped.TryGetValue(key, out var list))
                grouped[key] = list = new List<Observation>();
            list.Add(obs);
        }

        foreach (var (key, group) in grouped)
        {
            var values = group.Select(o => o.Points).Distinct().OrderBy(v => v).ToList();
            scores[key] = values.Max();
            if (values.Count > 1)
            {
                var (memberId, day) = key;
                issues.Add(
                    $"Conflicting duplicate scores for member {memberId} on {day}: " +
                    string.Join(", ", values.Select(v => v.ToString("N0"))) + ". Highest value used.");
            }
        }

        var observedDays = observations
            .Select(o => o.Day)
            .Distinct()
            .OrderBy(d => Array.IndexOf(TextUtil.DayOrder, d) is var i && i >= 0 ? i : 99)
            .ToList();

        var missingByDay = new Dictionary<string, List<Member>>();
        foreach (var day in observedDays)
        {
            var matchedIds = scores.Where(kv => kv.Key.Item2 == day).Select(kv => kv.Key.Item1).ToHashSet();
            missingByDay[day] = members.Where(m => !matchedIds.Contains(m.MemberId)).ToList();
        }

        return new WeeklyData(observations, scores, issues, missingByDay);
    }
}
